internal/testrunner/runners/system/runner.go

```diff
@@ -130,10 +138,172 @@ func (r *runner) TestFolderRequired() bool {
 	return true
 }
 
+// CanRunSetupTeardownIndependent returns whether this test runner can run setup or
+// teardown process independent.
+func (r *runner) CanRunSetupTeardownIndependent() bool {
+	return true
+}
+
+func createSetupServicesDir(elasticPackagePath *locations.LocationManager) error {
+	dirPath := elasticPackagePath.ServiceSetupDir()
+	err := os.MkdirAll(dirPath, 0755)
+	if err != nil {
+		return fmt.Errorf("mkdir failed (path: %s): %w", dirPath, err)
+	}
+	return nil
+}
+
 // Run runs the system tests defined under the given folder
 func (r *runner) Run(options testrunner.TestOptions) ([]testrunner.TestResult, error) {
 	r.options = options
-	return r.run()
+
+	if !r.options.RunSetup && !r.options.RunTearDown && !r.options.RunTestsOnly {
+		return r.run()
+	}
+
+	result := r.newResult("(init)")
+	if err := r.initRun(); err != nil {
+		return result.WithError(err)
+	}
+
+	if r.options.RunSetup {
+		// variant information in runTestOnly or runTearDown modes is retrieved from serviceOptions (file in setup dir)
+		if len(r.variants) > 1 {
+			return result.WithError(fmt.Errorf("a variant must be selected or trigger the test in no-variant mode (available variants: %s)", strings.Join(r.variants, ", ")))
+		}
+		if len(r.variants) == 0 {
+			logger.Debug("No variant mode")
+		}
+	}
+
+	_, err := os.Stat(r.locationManager.ServiceSetupDir())
+	logger.Debugf("Service setup data exists in %s: %v", r.locationManager.ServiceSetupDir(), !os.IsNotExist(err))
+	if r.options.RunSetup && !os.IsNotExist(err) {
+		return result.WithError(fmt.Errorf("failed to run --setup, required to tear down previous setup"))
+	}
```

Comment:
Nit. It'd be nice to make the setup more aware of the state of the current setup, so if the file exists:
* It checks if the setup is actually up, if not, it ignores existing file.
* It checks if the setup was started with the same parameters, then it just finishes, or runs idempotently.
* If the setup is running, and it has different parameters, then we show the current error.

Or maybe we can just try to always run it and trust on operations being idempotent.

But this would add even more complexity, so lets leave this for a future change unless you see a simple and reliable way to do it.

Reply:
Yes, this would mean that a sequence of commands like the one above would work. But agree to leave it for a future change.